java/code/src/com/suse/manager/maintenance/MaintenanceManager.java

```diff
@@ -93,6 +97,66 @@ public static MaintenanceManager instance() {
         return instance;
     }
 
+    /**
+     * Check if an action of given type can be scheduled at given date for given systems.
+     * If some systems have a {@link MaintenanceSchedule} and are outside of their maintenance windows,
+     * throw the {@link NotInMaintenanceModeException} that bears the offending schedules.
+     *
+     * @param systemIds the system IDs to check
+     * @param date the schedule date of the action
+     * @param actionType the type of action
+     * @throws NotInMaintenanceModeException when some systems are outside of maintenance window
+     */
+    public void checkMaintenanceWindows(Set<Long> systemIds, Date date, ActionType actionType) {
+        if (actionType.isMaintenancemodeOnly()) {
+            Set<MaintenanceSchedule> offendingSchedules = listSystemSchedulesNotMachingDate(systemIds, date);
+            if (!offendingSchedules.isEmpty()) {
+                throw new NotInMaintenanceModeException(offendingSchedules, date);
```

Comment:
In most cases "Fault" exceptions are only thrown from inside of XMLRPC Handler classes. There are a few
exceptions, but it seems some where from "us" and others have a special case where "Command" classes which may be used everywhere throw such exceptions.

On the other hand I do not see a big benefit in writing 2 Exceptions and convert one into the other.
Maybe we need to discuss it with the team?

Reply:
I know that this is not kosher, but I did this as a kind of compromise.  By declaring the exception as a `Fault` I don't need to care about the `Exception` -> `Fault` translation on the Handler side. Not doing this, I'd need to visit all the methods, that have something to do with action scheduling (in multiple handlers). Moreover, when somebody adds a xmlrpc method that schedules an action in the future, he/she will need to take care of proper translation. I think declaring that exception as a `Fault` is a small price to pay for the benefits. I also couldn't think of more drawbacks of doing this.

Reply:
yes, this is a good point.

Reply:
So, after the discussion, I removed the dependency on the fault from the backend. This makes total sense. On top of that, I implemented handling of `NotInMaintModeException` in a general sense, so that it doesn't need to be handled case-by-case.